ramips: add encrypted SGE factory image for D-Link devices

Lucky1openwrt opened the pull request:
creates SGE encrypted factory images
to use via the D-Link web interface
rename the old factory unencrypted images to recovery
for use in the recovery console when recovery is needed
DIR-1935-A1 , DIR-853-A1 , DIR-853-A3 , DIR-867-A1 ,
DIR-878-A1 and DIR-882-A1

aparcar commented:
I'm unsure about the naming here, I'd prefer we have factory.bin and recovery.bin, any other opinions?

Lucky1openwrt commented:
there was a discussion about this here 
https://github.com/openwrt/openwrt/pull/3109
https://forum.openwrt.org/t/add-support-for-d-link-covr-x1860/161862/26

musashino205 commented:
IMO: ```ramips: add factory images for D-Link SGE devices```

musashino205 commented:
IMO: ```ramips: add SGE factory image for D-Link devices```

Lucky1openwrt commented:
I'm happy to change the comments to anything
just has to make sense and as we had images call factory
maybe 
ramips: add encrypted SGE factory image for D-Link devices

s-2 commented:
As a side note, there seems to be much interest in these devices (or at least COVR-X1860, using the same `factory` / `recovery` image formats) for use with gluon.
While I can't recall in detail how the image name extensions work over there, separate image types exist for `factory` and `other` (which could be recovery here), at least for factory there is an option to specify a different image file extension (`factory_ext`). Maybe this also affects the naming of the recovery image, so that maybe we might want to aim for `factory-recovery.bin` rather than just `recovery`?
But technically, that's still just a downstream issue OpenWrt should not be required to deal with.

Regarding the change in `$(Device/dlink_dir-8xx-a1)`, it is confirmed that all devices that include this recipe use the same keys (i.e. factory images decrypt and verify correctly)? I could test some, but don't own all from the series.
For some devices (e.g. DIR-2660 / 3060) the keys are not yet known, but those use a different include, so that should be fine.

Regarding the relocation of the include, I think this is becasue the `DEVICE_MODEL` variable was not found otherwise?
We can probably use double escapes here, i.e. use `$$$$(DEVICE_MODEL)` instead of `$$(DEVICE_MODEL)`, so the include could remain at the top of each device's recipe.

Lucky1openwrt commented:
just changed if from factory-recovery and factory-webflash
I'm not sure how the "firmware selector" will work with this
I don't like the fact that there was a factory image & now there still is but it's not the same 
I'm happy to change anything just want it to go through in some form
all the devises listed are the NOR version & use the same key
I have tested all of them but uploading the factory firmware for each model
into my DIR-878-A1 and then using new factory file to openwrt
I have an update waiting with charges the NAND factory to recovery
and added the DIR-853-A3 but I can't test this as I don't have the OEM firmware
the others I think are all private keys we don't have
and yes I move DEVICE_MODEL but can change if needed if it will work

Lucky1openwrt commented:
* DIR-853-A3 needs testing

s-2 commented:
> I don't like the fact that there was a factory image & now there still is but it's not the same

Same here, not sure what would be the ideal solution (also regarding other devices).
I did not even think of firmware-selector, sounds like the issue is quite similar to the gluon firmware selector / autoupdater 🤔 (but that would have to be considered downstream).

raenye commented:
(off-topic)
@s-2 the alpha factory image encryption program from #3109 never made it to OpenWrt, right? it's needed for DAP-1720 as well.

s-2 commented:
> the alpha factory image encryption program from #3109 never made it to OpenWrt, right? it's needed for DAP-1720 as well.

correct, there has been no additional attempt since then. It is also needed for [DIR-X1860](https://forum.openwrt.org/t/adding-openwrt-support-for-d-link-dir-x1860-mt7621-mt7915-ax1800/106500).

raenye commented:
> I had closed the PR after 1.5 years of continuous rebasing

I rebased it recently, is it ok with you if I resubmit it as a new PR? There was also a related SEAMA seal fix IIRC.

raenye commented:
@Lucky1openwrt perhaps mention SGE also in the commit body text, since D-link uses multiple ODMs.

raenye commented on target/linux/ramips/image/mt7621.mk:
```diff
@@ -684,10 +684,12 @@ define Device/dlink_dir-8xx-a1
 	check-size | append-metadata
-  IMAGE/factory.bin := append-kernel | append-rootfs | check-size
+  IMAGE/recovery.bin := append-kernel | append-rootfs | check-size
+  IMAGE/factory.bin := append-kernel | append-rootfs | check-size | \
```
Why not start from IMAGE/recovery.bin?

s-2 replied:
makes sense, at least for devices with this image format it could be generated incrementally

Lucky1openwrt replied:
so change it to 
IMAGE/factory.bin := $$(IMAGE/recovery.bin) | dlink-sge-image $$$$(DEVICE_MODEL)
?

raenye replied:
I think this should work.

raenye replied:
BTW, since we have the following in `target/linux/ramips/image/Makefile`:
```
define Device/Default
  [...]
  IMAGES := sysupgrade.bin
  COMPILE :=
  sysupgrade_bin := append-kernel | append-rootfs | pad-rootfs
  IMAGE/sysupgrade.bin := append-kernel | append-rootfs | pad-rootfs | check-size | append-metadata
endef
```
You could drop the `IMAGE/sysupgrade.bin` line.

Lucky1openwrt replied:
ok Dropped

raenye commented on target/linux/ramips/image/mt7621.mk:
```diff
@@ -760,6 +762,9 @@ define Device/dlink_dir-853-a3
   DEVICE_VARIANT := A3
+  IMAGES += factory.bin
+  IMAGE/factory.bin := append-kernel | pad-to $$(KERNEL_SIZE) | \
+	append-ubi | check-size | dlink-sge-image $$(DEVICE_MODEL)
```
Why not start from IMAGE/recovery.bin?

s-2 commented:
> is it ok with you if I resubmit it as a new PR? There was also a related SEAMA seal fix IIRC.

Sure, the SEAMA seal fix was done by someone else though.
The only thing I did there was allowing to use a parameter for the board signature, since DIR-842 C3 (EU) needs a different `SEAMA_SIGNATURE` for the recovery than for the encrypted factory image, so it needed to be passed as a parameter, rather than a static device variable.
Please also add DIR-X1860 (and DAP-1720 of course) to the commit description (AES keys would be introduced in the image recipes later, since they all happen to be alphanumeric).

By the way, there is also a tool for [M32 and R32](https://github.com/openwrt/openwrt/pull/13141) written by @RolandoMagico 
(maybe he wants to open a PR as well, while we're at it 🙂)

raenye commented:
> Please also add DIR-X1860 (and DAP-1720 of course) to the commit description

Thanks, I will.
Besides rebasing and editing the commit message I did nothing, so not sure what should be the tags. Authored-by you, Signed-off-by me? or Signed-off-by both?

s-2 commented:
> Besides rebasing and editing the commit message

While thinking about how old this thing is, you probably see deprecation warnings from OpenSSL already?
This should indeed be migrated to the new EVP API first (I had tried initially back then, but found it way to complicated, so sticked with the old API). For `dlink-sge-image` I had migrated everything to the new API (since there was so much interest in the COVR-X1860 suddenly), and then again one step back, since my local OpenSSL was too new... (so don't use the `_ex2` methods in OpenWrt yet, just the legacy `_ex`).
Maybe I really should have another look at it again if I find the time and proceed with migrating, not sure if there's still much interest in DIR-842 / DIR-850 after three years, and even DIR-X1860 has been EOL for a while, being the first ever AX device from D-Link.

Lucky1openwrt committed:
ramips: add encrypted SGE factory image for D-Link devices

creates SGE encrypted factory images
to use via the D-Link web interface
rename the old factory unencrypted images to recovery
for use in the recovery console when recovery is needed
DIR-1935-A1 , DIR-853-A1 , DIR-853-A3 , DIR-867-A1 ,
DIR-878-A1 and DIR-882-A1

Signed-off-by: Alan Luck <luckyhome2008@gmail.com>

raenye commented:
(still O/T)

> Maybe I really should have another look at it again if I find the time and proceed with migrating,

I think it's possible to get rid of the firmware-utils addition altogether and just code everything in `generic.mk`. Within `Build/alpha_encimg` invoke `openssl aes-256-cbc -e` with the mangled key and iv hard-coded within the `Device/dlink_xxx` stanzas. Since we anyway have hard-coded key&iv, they may as well be after the xor with its seama signature. Adding support for a new device would just require mangling the key&iv once, so there should be a script/documentation for this.

s-2 commented:
> with the mangled key and iv hard-coded

(fun fact: there is `nec-enc.c` that generates exactly this xor pattern for full image files; those NEC devices were built by Alpha Networks, so the format is like a predecessor of alpha_encimg, i.e. now they are just using this for the keys, no longer the full image. Maybe we should really have a new forum/thread / PR for those devices)

raenye commented:
> (fun fact: there is `nec-enc.c`  [...] Maybe we should really have a new forum/thread / PR for those devices)

Thank you for this reference. I'll have a look, prepare a draft PR (already started something) and we'll continue the discussion there.
Sorry everybody for littering the conversation.

RolandoMagico commented:
> maybe he wants to open a PR as well, while we're at it

Sure, but sometimes encryption of images is not working as expected when using it in the OpenWrt build, I’m still investigating what’s wrong there

hauke commented:
Thank you for the PR, I merged it into main in 4c0dc68f466588bb2ef60b29a4d8052b34112ca6.
